File: apps/desktop/src/utils/waf.ts
const DEFAULT_WAF_SUFFIXES = ["bmp", "gif", "jpeg", "jpg", "png", "svg", "webp"] as const;

const toSortedUnique = (values: Iterable<string>): string[] =>
  Array.from(new Set(values)).sort((left, right) => left.localeCompare(right));

const containsWafBypassTokens = (value: string): boolean => {
  for (const character of value) {
    const codePoint = character.codePointAt(0) ?? 0;
    if (
      codePoint <= 0x1f ||
      codePoint === 0x7f ||
      /\s/u.test(character) ||
      codePoint > 0x7f ||
      "^$;%?#=".includes(character)
    ) {
      return true;
    }
  }

  const normalized = value.replace(/\\/g, "/");
  return (
    normalized.includes("//") ||
    normalized.includes("/./") ||
    normalized.includes("/../") ||
    normalized.startsWith("./") ||
    normalized.startsWith("../") ||
    normalized.endsWith("/.") ||
    normalized.endsWith("/..") ||
    normalized === "." ||
    normalized === ".."
  );
};

export const normalizeWafSuffix = (value?: string | null): string | undefined => {
  if (!value) {
    return undefined;
  }

  const cleaned = value.trim().replace(/^\.+/, "");

  if (
    cleaned.length === 0 ||
    cleaned.length > 16 ||
    containsWafBypassTokens(cleaned) ||
    !/^[a-zA-Z0-9]+$/.test(cleaned)
  ) {
    return undefined;
  }

  return cleaned.toLowerCase();
};

export const extractSuffixFromFileName = (
  fileName?: string | null
): string | undefined => {
  if (!fileName) {
    return undefined;
  }

  const dotIndex = fileName.lastIndexOf(".");
  if (dotIndex < 0 || dotIndex === fileName.length - 1) {
    return undefined;
  }

  return normalizeWafSuffix(fileName.slice(dotIndex + 1));
};

export const extractSuffixFromObjectKey = (
  objectKey?: string | null
): string | undefined => {
  if (!objectKey) {
    return undefined;
  }

  const normalizedKey = normalizeWafObjectKey(objectKey);
  if (!normalizedKey) {
    return undefined;
  }

  const fileName = normalizedKey.split("/").pop() ?? normalizedKey;
  return extractSuffixFromFileName(fileName);
};

export const resolveOrderedWafSuffixes = (
  candidates: Array<string | undefined | null>
): string[] => {
  const normalized = toSortedUnique(
    candidates
      .map((value) => normalizeWafSuffix(value))
      .filter((value): value is string => Boolean(value))
  );

  if (normalized.length > 0) {
    return normalized;
  }

  return [...DEFAULT_WAF_SUFFIXES];
};

export const collectWafSuffixes = (
  fileNames: Array<string | undefined | null>,
  objectKeys: Array<string | undefined | null>
): string[] => {
  const fromObjectKeys = objectKeys
    .map((value) => extractSuffixFromObjectKey(value))
    .filter((value): value is string => Boolean(value));

  const fromFileNames = fileNames
    .map((value) => extractSuffixFromFileName(value))
    .filter((value): value is string => Boolean(value));

  return resolveOrderedWafSuffixes([...fromObjectKeys, ...fromFileNames]);
};

export const buildWafPattern = (
  digitCount: number,
  suffixes: Array<string | undefined | null>
): string => {
  const orderedSuffixes = resolveOrderedWafSuffixes(suffixes);
  const suffixGroup = orderedSuffixes.join("|");

  return String.raw`^/img/public/[0-9]{${digitCount}}\.(?:${suffixGroup})$`;
};

interface WafCdnScope {
  scheme?: string;
  host?: string;
  pathPrefix: string;
}

interface WafObjectKeyParts {
  prefix: string;
  number: string;
  suffix: string;
}

interface WafObjectGroup {
  prefix: string;
  suffix: string;
  width: number;
  numbers: string[];
}

const wafRegexMeta = /[.*+?^${}()|[\]\\]/g;

const escapeWafRegex = (value: string): string => value.replace(wafRegexMeta, "\\$&");

const buildDigitRange = (start: number, end: number): string =>
  start === end ? `${start}` : `[${start}-${end}]`;

const parseWafObjectKeyParts = (
  value?: string | null,
  pathPrefix = ""
): WafObjectKeyParts | undefined => {
  const normalized = normalizeWafObjectKey(value, pathPrefix);
  if (!normalized) {
    return undefined;
  }

  const lastSlash = normalized.lastIndexOf("/");
  const lastDot = normalized.lastIndexOf(".");
  if (lastSlash < 0 || lastDot <= lastSlash || lastDot === normalized.length - 1) {
    return undefined;
  }

  const number = normalized.slice(lastSlash + 1, lastDot);
  const suffix = normalized.slice(lastDot + 1);

  if (!/^\d+$/.test(number)) {
    return undefined;
  }

  return {
    prefix: normalized.slice(0, lastSlash + 1),
    number,
    suffix
  };
};

const buildNumericRangeRegex = (start: string, end: string): string => {
  if (start === end) {
    return start;
  }

  let prefixLength = 0;
  while (prefixLength < start.length && start[prefixLength] === end[prefixLength]) {
    prefixLength += 1;
  }

  const prefix = start.slice(0, prefixLength);
  const startDigit = Number(start[prefixLength]);
  const endDigit = Number(end[prefixLength]);
  const remainingLength = start.length - prefixLength - 1;

  if (remainingLength === 0) {
    return `${prefix}${buildDigitRange(startDigit, endDigit)}`;
  }

  const parts = [
    buildNumericRangeRegex(start, `${prefix}${start[prefixLength]}${"9".repeat(remainingLength)}`)
  ];

  if (startDigit + 1 <= endDigit - 1) {
    parts.push(
      `${prefix}${buildDigitRange(startDigit + 1, endDigit - 1)}[0-9]{${remainingLength}}`
    );
  }

  parts.push(
    buildNumericRangeRegex(
      `${prefix}${end[prefixLength]}${"0".repeat(remainingLength)}`,
      end
    )
  );

  const uniqueParts = Array.from(new Set(parts));
  return uniqueParts.length === 1 ? uniqueParts[0] : `(?:${uniqueParts.join("|")})`;
};

const normalizeWafPathPrefix = (value: string): string => {
  const cleaned = value
    .trim()
    .replace(/\\/g, "/")
    .replace(/\/+$/, "");

  if (cleaned.length === 0 || cleaned === "/") {
    return "";
  }

  return cleaned.startsWith("/") ? cleaned : `/${cleaned}`;
};

const buildWafGuardedPathPrefix = (pathPrefix: string): string =>
  pathPrefix ? `${pathPrefix}/img/public/` : "/img/public/";

const resolveWafCdnScope = (cdnBaseUrl?: string | null): WafCdnScope => {
  const raw = cdnBaseUrl?.trim();
  if (!raw) {
    return { pathPrefix: "" };
  }

  try {
    const parsed = new URL(raw);
    return {
      scheme: parsed.protocol.replace(/:$/, "").toLowerCase(),
      host: parsed.hostname.toLowerCase(),
      pathPrefix: normalizeWafPathPrefix(parsed.pathname)
    };
  } catch {
    return { pathPrefix: "" };
  }
};

const buildWafFullUriPrefix = (scope: WafCdnScope): string | undefined => {
  if (!scope.scheme || !scope.host) {
    return undefined;
  }

  return `${scope.scheme}://${scope.host}${buildWafGuardedPathPrefix(scope.pathPrefix)}`;
};

const isStrictAllowlistPath = (path: string, guardedPrefix: string): boolean => {
  if (!path.startsWith(guardedPrefix)) {
    return false;
  }

  const filePart = path.slice(guardedPrefix.length);
  if (filePart.length === 0 || filePart.includes("/") || containsWafBypassTokens(filePart)) {
    return false;
  }

  const dotIndex = filePart.lastIndexOf(".");
  if (dotIndex <= 0 || dotIndex === filePart.length - 1) {
    return false;
  }

  const number = filePart.slice(0, dotIndex);
  const suffix = filePart.slice(dotIndex + 1);

  if (!/^\d+$/.test(number) || number.length > 20) {
    return false;
  }

  if (!/^[a-zA-Z0-9]+$/.test(suffix) || suffix.length > 16) {
    return false;
  }

  return true;
};

const buildWafSetLiteral = (values: string[]): string =>
  `{${values.map((value) => quoteWafString(value)).join(" ")}}`;

const buildWafFullUri = (scope: WafCdnScope, normalizedPath: string): string =>
  `${scope.scheme}://${scope.host}${normalizedPath}`;

const collectWafAllowlistFullUris = (
  objectKeys: Array<string | undefined | null>,
  cdnBaseUrl?: string | null
): string[] => {
  const scope = resolveWafCdnScope(cdnBaseUrl);
  const guardedPrefix = buildWafGuardedPathPrefix(scope.pathPrefix);
  const uniqueUris = new Set<string>();

  for (const value of objectKeys) {
    const normalized = normalizeWafObjectKey(value, scope.pathPrefix);
    if (normalized && isStrictAllowlistPath(normalized, guardedPrefix) && scope.scheme && scope.host) {
      uniqueUris.add(buildWafFullUri(scope, normalized));
    }
  }

  return toSortedUnique(uniqueUris);
};

const collectWafAllowlistPaths = (
  objectKeys: Array<string | undefined | null>,
  cdnBaseUrl?: string | null
): string[] => {
  const scope = resolveWafCdnScope(cdnBaseUrl);
  const guardedPrefix = buildWafGuardedPathPrefix(scope.pathPrefix);
  const uniquePaths = new Set<string>();

  for (const value of objectKeys) {
    const normalized = normalizeWafObjectKey(value, scope.pathPrefix);
    if (normalized && isStrictAllowlistPath(normalized, guardedPrefix)) {
      uniquePaths.add(normalized);
    }
  }

  return toSortedUnique(uniquePaths);
};

const normalizeWafObjectKey = (
  value?: string | null,
  pathPrefix = ""
): string | undefined => {
  if (!value) {
    return undefined;
  }

  const trimmed = value.trim();
  if (!trimmed || containsWafBypassTokens(trimmed)) {
    return undefined;
  }

  const cleaned = trimmed
    .replace(/\\/g, "/")
    .replace(/^\/+/, "");

  if (cleaned.length === 0 || containsWafBypassTokens(cleaned)) {
    return undefined;
  }

  const normalizedPath = `/${cleaned}`;
  if (!pathPrefix) {
    return normalizedPath;
  }

  return `${pathPrefix}${normalizedPath}`;
};

const quoteWafString = (value: string): string => JSON.stringify(value);

const collectWafObjectFragments = (
  objectKeys: Array<string | undefined | null>,
  cdnBaseUrl?: string | null
): string[] => {
  const scope = resolveWafCdnScope(cdnBaseUrl);
  const grouped = new Map<string, WafObjectGroup>();

  for (const value of objectKeys) {
    const parts = parseWafObjectKeyParts(value, scope.pathPrefix);
    if (!parts) {
      continue;
    }

    const groupKey = `${parts.prefix}\u0000${parts.suffix}\u0000${parts.number.length}`;
    const existing = grouped.get(groupKey);
    if (existing) {
      existing.numbers.push(parts.number);
      continue;
    }

    grouped.set(groupKey, {
      prefix: parts.prefix,
      suffix: parts.suffix,
      width: parts.number.length,
      numbers: [parts.number]
    });
  }

  const fragments: string[] = [];

  for (const group of grouped.values()) {
    const orderedNumbers = toSortedUnique(group.numbers);
    if (orderedNumbers.length === 0) {
      continue;
    }

    let runStart = orderedNumbers[0];
    let previous = orderedNumbers[0];

    for (const current of orderedNumbers.slice(1)) {
      if (BigInt(current) === BigInt(previous) + 1n) {
        previous = current;
        continue;
      }

      fragments.push(
        `${escapeWafRegex(group.prefix)}${buildNumericRangeRegex(runStart, previous)}\\.${escapeWafRegex(group.suffix)}`
      );
      runStart = current;
      previous = current;
    }

    fragments.push(
      `${escapeWafRegex(group.prefix)}${buildNumericRangeRegex(runStart, previous)}\\.${escapeWafRegex(group.suffix)}`
    );
  }

  fragments.sort();
  return fragments;
};

export const buildWafObjectPattern = (
  objectKeys: Array<string | undefined | null>,
  cdnBaseUrl?: string | null
): string => {
  const fragments = collectWafObjectFragments(objectKeys, cdnBaseUrl);

  if (fragments.length === 0) {
    return "";
  }

  if (fragments.length === 1) {
    return `^${fragments[0]}$`;
  }

  return `^(?:${fragments.join("|")})$`;
};

export const buildWafObjectExpression = (
  objectKeys: Array<string | undefined | null>,
  cdnBaseUrl?: string | null
): string => {
  const scope = resolveWafCdnScope(cdnBaseUrl);
  const allowlistedPaths = collectWafAllowlistPaths(objectKeys, cdnBaseUrl);

  if (scope.host) {
    const hostCondition = `http.host eq ${quoteWafString(scope.host)}`;

    if (allowlistedPaths.length === 0) {
      return hostCondition;
    }

    return `${hostCondition} and not (raw.http.request.uri.path in ${buildWafSetLiteral(allowlistedPaths)})`;
  }

  const pathPrefix = buildWafGuardedPathPrefix(scope.pathPrefix);
  const pathCondition = `starts_with(http.request.uri.path, ${quoteWafString(pathPrefix)})`;

  if (allowlistedPaths.length === 0) {
    return pathCondition;
  }

  return `${pathCondition} and not (raw.http.request.uri.path in ${buildWafSetLiteral(allowlistedPaths)})`;
};

export const defaultWafSuffixes = [...DEFAULT_WAF_SUFFIXES];

File: src-tauri/src/runtime/adapter_runtime/waf.rs
//! WAF allowlist and expression builder.
//!
//! `AdapterRuntime` delegates WAF scope parsing, allowlist filtering, and
//! fingerprint generation here. The resulting expression is consumed by the
//! Cloudflare transport module when syncing zone rules.

use super::WafScope;
use sha2::{Digest, Sha256};
use std::collections::BTreeSet;

#[cfg(test)]
use std::collections::BTreeMap;

pub(super) fn resolve_waf_scope(cdn_base_url: &str) -> Option<WafScope> {
  let parsed = reqwest::Url::parse(cdn_base_url.trim()).ok()?;
  let scheme = parsed.scheme().to_ascii_lowercase();
  let host = parsed.host_str()?.to_ascii_lowercase();
  let path_prefix = normalize_waf_path_prefix(parsed.path());

  Some(WafScope {
    scheme,
    host,
    path_prefix,
  })
}

fn normalize_waf_path_prefix(value: &str) -> String {
  let cleaned = value
    .trim()
    .replace('\\', "/")
    .trim_end_matches('/')
    .to_string();

  if cleaned.is_empty() || cleaned == "/" {
    return String::new();
  }

  if cleaned.starts_with('/') {
    cleaned
  } else {
    format!("/{cleaned}")
  }
}

fn build_waf_guarded_path_prefix(path_prefix: &str) -> String {
  if path_prefix.is_empty() {
    "/img/public/".to_string()
  } else {
    format!("{path_prefix}/img/public/")
  }
}

pub(super) fn build_object_allowlist_expression(object_keys: &[String], scope: &WafScope) -> String {
  let host_condition = format!("http.host eq {}", quote_waf_string(scope.host.as_str()));
  let allowlisted_paths = collect_waf_allowlist_paths(object_keys, scope.path_prefix.as_str());

  if allowlisted_paths.is_empty() {
    return host_condition;
  }

  format!(
    "{host_condition} and not (raw.http.request.uri.path in {})",
    build_waf_set_literal(allowlisted_paths.as_slice())
  )
}

pub(super) fn collect_waf_allowlist_paths(object_keys: &[String], path_prefix: &str) -> Vec<String> {
  let mut unique_uris = BTreeSet::new();
  let guarded_prefix = build_waf_guarded_path_prefix(path_prefix);

  for object_key in object_keys {
    if let Some(normalized) = normalize_waf_object_key(object_key.as_str(), path_prefix) {
      if is_strict_allowlist_path(normalized.as_str(), guarded_prefix.as_str()) {
        unique_uris.insert(normalized);
      }
    }
  }

  unique_uris.into_iter().collect::<Vec<_>>()
}

fn build_waf_set_literal(values: &[String]) -> String {
  let quoted_values = values
    .iter()
    .map(|value| quote_waf_string(value.as_str()))
    .collect::<Vec<_>>();

  format!("{{{}}}", quoted_values.join(" "))
}

fn contains_waf_bypass_tokens(value: &str) -> bool {
  if value.chars().any(|ch| {
    ch.is_ascii_control()
      || ch.is_whitespace()
      || !ch.is_ascii()
      || matches!(ch, '^' | '$' | ';' | '%' | '?' | '#' | '=')
  }) {
    return true;
  }

  let normalized = value.replace('\\', "/");
  normalized.contains("//")
    || normalized.contains("/./")
    || normalized.contains("/../")
    || normalized.starts_with("./")
    || normalized.starts_with("../")
    || normalized.ends_with("/.")
    || normalized.ends_with("/..")
    || normalized == "."
    || normalized == ".."
}

fn is_strict_allowlist_path(path: &str, guarded_prefix: &str) -> bool {
  if !path.starts_with(guarded_prefix) {
    return false;
  }

  let file_part = &path[guarded_prefix.len()..];
  if file_part.is_empty() || file_part.contains('/') || contains_waf_bypass_tokens(file_part) {
    return false;
  }

  let Some((number, suffix)) = file_part.rsplit_once('.') else {
    return false;
  };

  if number.is_empty() || number.len() > 20 || !number.chars().all(|ch| ch.is_ascii_digit()) {
    return false;
  }

  if suffix.is_empty()
    || suffix.len() > 16
    || !suffix.chars().all(|ch| ch.is_ascii_alphanumeric())
  {
    return false;
  }

  true
}

pub(super) fn hash_waf_allowlist_values(values: &[String]) -> String {
  let mut ordered_values = values.to_vec();
  ordered_values.sort();
  ordered_values.dedup();

  let mut hasher = Sha256::new();
  for value in ordered_values {
    hasher.update(value.as_bytes());
    hasher.update([0]);
  }

  format!("{:x}", hasher.finalize())
}

fn normalize_waf_object_key(value: &str, path_prefix: &str) -> Option<String> {
  let trimmed = value.trim();
  if trimmed.is_empty() || contains_waf_bypass_tokens(trimmed) {
    return None;
  }

  let cleaned = trimmed
    .replace('\\', "/")
    .trim_start_matches('/')
    .to_string();

  if cleaned.is_empty() || contains_waf_bypass_tokens(cleaned.as_str()) {
    None
  } else {
    let normalized_path = format!("/{cleaned}");
    if path_prefix.is_empty() {
      Some(normalized_path)
    } else {
      Some(format!("{path_prefix}{normalized_path}"))
    }
  }
}

fn quote_waf_string(value: &str) -> String {
  serde_json::to_string(value).unwrap_or_else(|_| {
    let escaped = value
      .replace('\\', "\\\\")
      .replace('"', "\\\"");
    format!("\"{escaped}\"")
  })
}

#[cfg(test)]
#[derive(Clone, Debug)]
struct WafObjectGroup {
  prefix: String,
  suffix: String,
  numbers: Vec<String>,
}

#[cfg(test)]
fn escape_waf_regex(value: &str) -> String {
  value
    .chars()
    .fold(String::new(), |mut output, ch| {
      if matches!(
        ch,
        '.' | '+' | '*' | '?' | '^' | '$' | '{' | '}' | '(' | ')' | '[' | ']' | '|' | '\\'
      ) {
        output.push('\\');
      }
      output.push(ch);
      output
    })
}

#[cfg(test)]
fn build_digit_range(start: u32, end: u32) -> String {
  if start == end {
    start.to_string()
  } else {
    format!("[{start}-{end}]")
  }
}

#[cfg(test)]
fn parse_waf_object_key_parts(value: &str, path_prefix: &str) -> Option<(String, String, String)> {
  let normalized = normalize_waf_object_key(value, path_prefix)?;
  let last_slash = normalized.rfind('/')?;
  let last_dot = normalized.rfind('.')?;

  if last_dot <= last_slash || last_dot + 1 >= normalized.len() {
    return None;
  }

  let number = &normalized[last_slash + 1..last_dot];
  if !number.chars().all(|ch| ch.is_ascii_digit()) {
    return None;
  }

  let prefix = normalized[..last_slash + 1].to_string();
  let suffix = normalized[last_dot + 1..].to_string();
  Some((prefix, number.to_string(), suffix))
}

#[cfg(test)]
fn build_numeric_range_regex(start: &str, end: &str) -> String {
  if start == end {
    return start.to_string();
  }

  let start_bytes = start.as_bytes();
  let end_bytes = end.as_bytes();
  let mut prefix_len = 0;
  while prefix_len < start_bytes.len() && start_bytes[prefix_len] == end_bytes[prefix_len] {
    prefix_len += 1;
  }

  let prefix = &start[..prefix_len];
  let start_digit = (start_bytes[prefix_len] - b'0') as u32;
  let end_digit = (end_bytes[prefix_len] - b'0') as u32;
  let remaining_len = start_bytes.len() - prefix_len - 1;

  if remaining_len == 0 {
    return format!("{prefix}{}", build_digit_range(start_digit, end_digit));
  }

  let mut parts = vec![build_numeric_range_regex(
    start,
    &format!("{prefix}{}{}", start_bytes[prefix_len] as char, "9".repeat(remaining_len)),
  )];

  if start_digit + 1 <= end_digit.saturating_sub(1) {
    parts.push(format!(
      "{prefix}{}[0-9]{{{remaining_len}}}",
      build_digit_range(start_digit + 1, end_digit - 1)
    ));
  }

  parts.push(build_numeric_range_regex(
    &format!("{prefix}{}{}", end_bytes[prefix_len] as char, "0".repeat(remaining_len)),
    end,
  ));

  parts.sort();
  parts.dedup();
  if parts.len() == 1 {
    parts.remove(0)
  } else {
    format!("(?:{})", parts.join("|"))
  }
}

#[cfg(test)]
fn collect_waf_object_fragments(object_keys: &[String], path_prefix: &str) -> Vec<String> {
  let mut grouped: BTreeMap<String, WafObjectGroup> = BTreeMap::new();

  for value in object_keys {
    let Some((prefix, number, suffix)) = parse_waf_object_key_parts(value.as_str(), path_prefix) else {
      continue;
    };

    let group_key = format!("{prefix}\0{suffix}\0{}", number.len());
    let entry = grouped.entry(group_key).or_insert_with(|| WafObjectGroup {
      prefix,
      suffix,
      numbers: Vec::new(),
    });
    entry.numbers.push(number);
  }

  let mut fragments = Vec::new();

  for group in grouped.values() {
    let mut ordered_numbers = group.numbers.clone();
    ordered_numbers.sort();
    ordered_numbers.dedup();

    if ordered_numbers.is_empty() {
      continue;
    }

    let mut run_start = ordered_numbers[0].clone();
    let mut previous = ordered_numbers[0].clone();

    for current in ordered_numbers.iter().skip(1) {
      let previous_value = previous
        .parse::<u128>()
        .expect("waf numbers should fit into u128");
      let current_value = current
        .parse::<u128>()
        .expect("waf numbers should fit into u128");

      if current_value == previous_value + 1 {
        previous = current.clone();
        continue;
      }

      fragments.push(format!(
        "{}{}\\.{}",
        escape_waf_regex(group.prefix.as_str()),
        build_numeric_range_regex(run_start.as_str(), previous.as_str()),
        escape_waf_regex(group.suffix.as_str())
      ));
      run_start = current.clone();
      previous = current.clone();
    }

    fragments.push(format!(
      "{}{}\\.{}",
      escape_waf_regex(group.prefix.as_str()),
      build_numeric_range_regex(run_start.as_str(), previous.as_str()),
      escape_waf_regex(group.suffix.as_str())
    ));
  }

  fragments.sort();
  fragments.dedup();
  fragments
}

#[cfg(test)]
fn build_waf_object_pattern(object_keys: &[String], path_prefix: &str) -> String {
  let fragments = collect_waf_object_fragments(object_keys, path_prefix);
  if fragments.is_empty() {
    return String::new();
  }

  if fragments.len() == 1 {
    return format!("^{}$", fragments[0]);
  }

  format!("^(?:{})$", fragments.join("|"))
}

#[cfg(test)]
mod tests {
  use super::{
    build_object_allowlist_expression,
    build_waf_object_pattern,
    hash_waf_allowlist_values,
    resolve_waf_scope,
  };

  #[test]
  fn blocks_all_public_paths_when_allowlist_is_empty() {
    let scope = resolve_waf_scope("https://cdn.example.com").expect("valid scope");
    let expression = build_object_allowlist_expression(&[], &scope);
    assert_eq!(expression, "http.host eq \"cdn.example.com\"");
  }

  #[test]
  fn builds_waf_object_allowlist_expression_with_exact_path_set() {
    let scope = resolve_waf_scope("https://cdn.example.com/assets/").expect("valid scope");
    let object_keys = [
      "img/public/000000002.webp".to_string(),
      "/img/public/000000001.png".to_string(),
      "img\\public\\000000003.jpg".to_string(),
    ];
    let expression = build_object_allowlist_expression(&object_keys, &scope);

    assert!(expression.contains("http.host eq \"cdn.example.com\""));
    assert_eq!(
      expression,
      "http.host eq \"cdn.example.com\" and not (raw.http.request.uri.path in {\"/assets/img/public/000000001.png\" \"/assets/img/public/000000002.webp\" \"/assets/img/public/000000003.jpg\"})"
    );
  }

  #[test]
  fn drops_non_strict_allowlist_entries_with_meta_chars() {
    let scope = resolve_waf_scope("https://cdn.example.com").expect("valid scope");
    let object_keys = [
      "img/public/000000001.png".to_string(),
      "img/public/000000002.^png".to_string(),
      "img/public/000000003.p$ng".to_string(),
      "img/public/000000006.p ng".to_string(),
      "img/public/000000007.p\u{00B7}ng".to_string(),
      "img/public/000000008.p\u{2026}ng".to_string(),
      "img/public/000000009.png=1".to_string(),
      "img/public/00A000004.png".to_string(),
      "img/public/000000005..png".to_string(),
    ];

    let expression = build_object_allowlist_expression(&object_keys, &scope);

    assert!(expression.contains("raw.http.request.uri.path in {"));
    assert!(expression.contains("\"/img/public/000000001.png\""));
    assert!(!expression.contains("\"https://cdn.example.com/img/public/000000002.^png\""));
    assert!(!expression.contains("\"https://cdn.example.com/img/public/000000003.p$ng\""));
    assert!(!expression.contains("\"https://cdn.example.com/img/public/000000006.p ng\""));
    assert!(!expression.contains("\"https://cdn.example.com/img/public/000000007.p\u{00B7}ng\""));
    assert!(!expression.contains("\"https://cdn.example.com/img/public/000000008.p\u{2026}ng\""));
    assert!(!expression.contains("\"https://cdn.example.com/img/public/000000009.png=1\""));
    assert!(!expression.contains("\"https://cdn.example.com/img/public/00A000004.png\""));
    assert!(!expression.contains("\"https://cdn.example.com/img/public/000000005..png\""));
  }

  #[test]
  fn drops_allowlist_entries_with_path_bypass_markers() {
    let scope = resolve_waf_scope("https://cdn.example.com").expect("valid scope");
    let object_keys = [
      "img/public/000000006.png".to_string(),
      "/img/public/../img/public/000000006.png".to_string(),
      "/img//img/public/000000006.png".to_string(),
      "/img/public/000000006.png;a=1".to_string(),
      "/img/public/;x=1/000000006.png".to_string(),
      "img/public/%2e%2e/etc/passwd".to_string(),
    ];

    let expression = build_object_allowlist_expression(&object_keys, &scope);

    assert!(expression.contains("raw.http.request.uri.path in {"));
    assert!(expression.contains("\"/img/public/000000006.png\""));
    assert!(!expression.contains("\"https://cdn.example.com/img/public/../img/public/000000006.png\""));
    assert!(!expression.contains("\"https://cdn.example.com/img//img/public/000000006.png\""));
    assert!(!expression.contains("\"https://cdn.example.com/img/public/000000006.png;a=1\""));
    assert!(!expression.contains("\"https://cdn.example.com/img/public/;x=1/000000006.png\""));
    assert!(!expression.contains("\"https://cdn.example.com/img/public/%2e%2e/etc/passwd\""));
  }

  #[test]
  fn blocks_all_when_allowlist_contains_only_non_strict_entries() {
    let scope = resolve_waf_scope("https://cdn.example.com").expect("valid scope");
    let object_keys = [
      "img/public/000000001.^png".to_string(),
      "img/public/000000002.p$ng".to_string(),
      "img/public/00A000003.png".to_string(),
    ];

    let expression = build_object_allowlist_expression(&object_keys, &scope);

    assert_eq!(expression, "http.host eq \"cdn.example.com\"");
    assert!(!expression.contains("not (http.request.full_uri in {"));
  }

  #[test]
  fn builds_waf_object_pattern_with_compressed_ranges() {
    let pattern = build_waf_object_pattern(&[
      "img/public/000000001.png".to_string(),
      "img/public/000000002.png".to_string(),
      "img/public/000000003.png".to_string(),
      "img/public/000000005.png".to_string(),
    ], "");

    assert_eq!(
      pattern,
      "^(?:/img/public/000000005\\.png|/img/public/00000000[1-3]\\.png)$"
    );
  }

  #[test]
  fn builds_waf_object_pattern_with_unique_sorted_paths() {
    let pattern = build_waf_object_pattern(&[
      "img/public/000000002.webp".to_string(),
      "/img/public/000000001.png".to_string(),
      "img/public/000000002.webp".to_string(),
    ], "/assets");

    assert_eq!(
      pattern,
      "^(?:/assets/img/public/000000001\\.png|/assets/img/public/000000002\\.webp)$"
    );
  }

  #[test]
  fn resolves_waf_scope_with_host_and_path_prefix() {
    let scope = resolve_waf_scope("https://cdn.example.com/media/v1/").expect("valid scope");
    assert_eq!(scope.scheme, "https");
    assert_eq!(scope.host, "cdn.example.com");
    assert_eq!(scope.path_prefix, "/media/v1");
  }

  #[test]
  fn fingerprints_allowlist_values_in_stable_order() {
    let a = hash_waf_allowlist_values(&[
      "https://cdn.example.com/img/public/000000001.png".to_string(),
      "https://cdn.example.com/img/public/000000002.webp".to_string(),
    ]);
    let b = hash_waf_allowlist_values(&[
      "https://cdn.example.com/img/public/000000002.webp".to_string(),
      "https://cdn.example.com/img/public/000000001.png".to_string(),
    ]);

    assert_eq!(a, b);
    assert_eq!(a.len(), 64);
  }
}